paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/NestedProjectedRowData.java

```diff
@@ -0,0 +1,244 @@
+        this.lastProjectedFields = new int[projectedFields.length];
+        for (int i = 0; i < projectedFields.length; i++) {
+            this.lastProjectedFields[i] = projectedFields[i][projectedFields[i].length - 1];
+        }
+
+        this.cachedFields = new Object[projectedFields.length];
+        this.isFieldsCached = new boolean[projectedFields.length];
+        this.cachedNullAt = new boolean[projectedFields.length];
+        this.isNullAtCached = new boolean[projectedFields.length];
+    }
+
+    public NestedProjectedRowData replaceRow(RowData row) {
+        this.row = row;
+        Arrays.fill(isFieldsCached, false);
+        Arrays.fill(isNullAtCached, false);
+        return this;
+    }
+
+    public static @Nullable NestedProjectedRowData copy(@Nullable NestedProjectedRowData rowData) {
+        if (rowData == null) {
+            return null;
+        }
+        return new NestedProjectedRowData(rowData.producedDataType, rowData.projectedFields);
+    }
+
+    @Override
+    public int getArity() {
+        return projectedFields.length;
+    }
+
+    @Override
+    public RowKind getRowKind() {
+        return row.getRowKind();
+    }
+
+    @Override
+    public void setRowKind(RowKind rowKind) {
+        row.setRowKind(rowKind);
+    }
+
+    @Override
+    public boolean isNullAt(int i) {
+        if (isNullAtCached[i]) {
+            return cachedNullAt[i];
+        }
+
+        RowData rowData = extractInternalRow(i);
+        boolean result;
+        if (rowData == null) {
+            result = true;
+        } else {
+            result = rowData.isNullAt(lastProjectedFields[i]);
+        }
+
+        isNullAtCached[i] = true;
+        cachedNullAt[i] = result;
+
+        return result;
+    }
+
+    @Override
+    public boolean getBoolean(int i) {
+        return getFieldAs(i, RowData::getBoolean);
+    }
+
+    @Override
+    public byte getByte(int i) {
+        return getFieldAs(i, RowData::getByte);
+    }
+
+    @Override
+    public short getShort(int i) {
+        return getFieldAs(i, RowData::getShort);
+    }
+
+    @Override
+    public int getInt(int i) {
+        return getFieldAs(i, RowData::getInt);
+    }
+
+    @Override
+    public long getLong(int i) {
+        return getFieldAs(i, RowData::getLong);
+    }
+
+    @Override
+    public float getFloat(int i) {
+        return getFieldAs(i, RowData::getFloat);
+    }
+
+    @Override
+    public double getDouble(int i) {
+        return getFieldAs(i, RowData::getDouble);
+    }
+
+    @Override
+    public StringData getString(int i) {
+        return getFieldAs(i, RowData::getString);
+    }
+
+    @Override
+    public DecimalData getDecimal(int i, int i1, int i2) {
+        return getFieldAs(i, (rowData, j) -> rowData.getDecimal(j, i1, i2));
+    }
+
+    @Override
+    public TimestampData getTimestamp(int i, int i1) {
+        return getFieldAs(i, (rowData, j) -> rowData.getTimestamp(j, i1));
+    }
```

Comment:
NIT: use meaningful argument name like `pos`, `precision`, `scale`. Don't use `i`, `i1` and `j`.